File: backend/app/services/schedule_generator.py
"""
Schedule Generator Service
Optimized algorithm with scoring, ranking, and required/optional course support.
"""

import itertools
import uuid
from typing import List, Dict, Optional, Set, Tuple
from datetime import time
from collections import defaultdict

from sqlalchemy.orm import Session

from app.models import Course, Section, Session as SessionModel, Schedule
from app.schemas import ScheduleResponse
from app.repositories.course_repository import CourseRepository
from app.repositories.section_repository import SectionRepository
from app.repositories.session_repository import SessionRepository


# ─── Constants ───────────────────────────────────────────────────────────────
DAY_ORDER = {
    # Spanish
    "lunes": 0, "martes": 1, "miércoles": 2, "miercoles": 2,
    "jueves": 3, "viernes": 4, "sábado": 5, "sabado": 5, "domingo": 6,
    # English
    "monday": 0, "tuesday": 1, "wednesday": 2,
    "thursday": 3, "friday": 4, "saturday": 5, "sunday": 6,
}

def _parse_day_to_idx(day_val) -> int:
    """Robustly parse a day string or integer to 0-6 index."""
    if day_val is None:
        return -1
        
    # If it's already an integer (e.g. 0-6 or 1-7)
    if isinstance(day_val, int):
        # Assume 1-7 or 0-6. If 1-7, shift down.
        if 1 <= day_val <= 7:
            return day_val - 1
        if 0 <= day_val <= 6:
            return day_val
        return -1

    day_str = str(day_val).strip().lower()
    if not day_str:
        return -1

    # Check if it's a numeric string
    if day_str.isdigit():
        val = int(day_str)
        if 1 <= val <= 7:
            return val - 1
        if 0 <= val <= 6:
            return val
        return -1

    # Exact match
    if day_str in DAY_ORDER:
        return DAY_ORDER[day_str]

    # Partial match (e.g. "lun" -> "lunes")
    for key, idx in DAY_ORDER.items():
        if key.startswith(day_str[:3]):  # match first 3 chars
            return idx

    return -1

SCORE_WEIGHTS = {
    "earliest": 0.30,
    "gaps": 0.30,
    "spread": 0.25,
    "compactness": 0.15,
}

MAX_MINUTES = 24 * 60  # normalizer


# ─── Time Slot Bitmap ────────────────────────────────────────────────────────
class TimeSlotBitmap:
    """
    Bitmap-based time-slot tracker for O(1) collision detection.
    Each day has slots in 5-minute granularity (288 slots per day).
    """

    SLOT_SIZE = 5  # minutes

    def __init__(self):
        # 7 days × 288 slots = 2016 bits, stored as a set for clarity
        self._occupied: Set[Tuple[int, int]] = set()

    def copy(self) -> "TimeSlotBitmap":
        new = TimeSlotBitmap()
        new._occupied = set(self._occupied)
        return new

    def add_session(self, day_idx: int, start_minutes: int, end_minutes: int) -> bool:
        """
        Try to add a session. Returns False if there's a collision.
        Does NOT modify state on collision.
        """
        if day_idx < 0 or day_idx > 6:
            return True  # unknown day → no collision

        slots_needed = set()
        t = start_minutes
        while t < end_minutes:
            slot = t // self.SLOT_SIZE
            key = (day_idx, slot)
            if key in self._occupied:
                return False  # collision
            slots_needed.add(key)
            t += self.SLOT_SIZE

        self._occupied.update(slots_needed)
        return True

    def can_add_session(self, day_idx: int, start_minutes: int, end_minutes: int) -> bool:
        """Check without mutating."""
        if day_idx < 0 or day_idx > 6:
            return True

        t = start_minutes
        while t < end_minutes:
            slot = t // self.SLOT_SIZE
            if (day_idx, slot) in self._occupied:
                return False
            t += self.SLOT_SIZE
        return True


# ─── Course Option ───────────────────────────────────────────────────────────
class CourseOption:
    """Represents a course section with its sessions, pre-computed for speed."""

    __slots__ = (
        "section_id", "course_id", "course_code", "course_name",
        "section_number", "professor", "sessions", "group",
        "_time_slots",
    )

    def __init__(self, section: Section, sessions: List[SessionModel], course: Course):
        self.section_id = section.id
        self.course_id = course.id
        self.course_code = course.code
        self.course_name = course.name
        self.section_number = section.section_number
        self.professor = section.professor
        self.sessions = sessions
        self.group = course.code

        # Pre-compute time slots: list of (day_idx, start_min, end_min)
        self._time_slots: List[Tuple[int, int, int]] = []
        for s in sessions:
            day_val = getattr(s, 'day', None) or getattr(s, 'day_of_week', None)
            day_idx = _parse_day_to_idx(day_val)
            start = getattr(s.start_time, 'hour', 0) * 60 + getattr(s.start_time, 'minute', 0)
            end = getattr(s.end_time, 'hour', 0) * 60 + getattr(s.end_time, 'minute', 0)
            self._time_slots.append((day_idx, start, end))

    def fits_bitmap(self, bitmap: TimeSlotBitmap) -> bool:
        """Check if this option can fit into the bitmap without collision."""
        for day_idx, start, end in self._time_slots:
            if not bitmap.can_add_session(day_idx, start, end):
                return False
        return True

    def add_to_bitmap(self, bitmap: TimeSlotBitmap) -> bool:
        """Add this option's sessions to bitmap. Returns False on collision."""
        for day_idx, start, end in self._time_slots:
            if not bitmap.add_session(day_idx, start, end):
                return False
        return True


# ─── Scoring ─────────────────────────────────────────────────────────────────
def score_combination(options: List[CourseOption]) -> Dict[str, float]:
    """
    Score a valid combination on 4 axes (0.0–1.0 each, higher = better).
    Returns individual scores + weighted composite.
    """
    if not options:
        return {"earliest": 0, "gaps": 0, "spread": 0, "compactness": 0, "composite": 0}

    # Collect all sessions across all options
    day_sessions: Dict[int, List[Tuple[int, int]]] = defaultdict(list)
    all_starts = []

    for opt in options:
        for day_idx, start, end in opt._time_slots:
            if day_idx >= 0:
                day_sessions[day_idx].append((start, end))
                all_starts.append(start)

    if not all_starts:
        return {"earliest": 0, "gaps": 0, "spread": 0, "compactness": 0, "composite": 0}

    # 1. Earliest score: lower average start = better
    avg_start = sum(all_starts) / len(all_starts)
    earliest = 1.0 - (avg_start / MAX_MINUTES)

    # 2. Gap score: fewer minutes of gaps between consecutive classes = better
    total_gap_minutes = 0
    for day_idx, intervals in day_sessions.items():
        sorted_intervals = sorted(intervals, key=lambda x: x[0])
        for i in range(1, len(sorted_intervals)):
            gap = sorted_intervals[i][0] - sorted_intervals[i - 1][1]
            if gap > 0:
                total_gap_minutes += gap

    # Normalize: 0 gaps = 1.0, 300+ minutes of gaps = 0.0
    gaps = max(0.0, 1.0 - (total_gap_minutes / 300.0))

    # 3. Spread score: how evenly distributed across days
    active_days = len(day_sessions)
    total_sessions = sum(len(v) for v in day_sessions.values())

    if active_days > 0 and total_sessions > 0:
        sessions_per_day = [len(day_sessions.get(d, [])) for d in range(7)]
        active_counts = [c for c in sessions_per_day if c > 0]
        if len(active_counts) > 1:
            avg_per_day = total_sessions / active_days
            variance = sum((c - avg_per_day) ** 2 for c in active_counts) / len(active_counts)
            spread = max(0.0, 1.0 - (variance / 4.0))  # variance of 4+ = 0
        else:
            spread = 0.0  # all on one day = bad spread
    else:
        spread = 0.0

    # 4. Compactness: fewer active days = better (less commute)
    compactness = max(0.0, 1.0 - ((active_days - 1) / 5.0))  # 1 day = 1.0, 6 days = 0.0

    composite = (
        SCORE_WEIGHTS["earliest"] * earliest
        + SCORE_WEIGHTS["gaps"] * gaps
        + SCORE_WEIGHTS["spread"] * spread
        + SCORE_WEIGHTS["compactness"] * compactness
    )

    return {
        "earliest": round(earliest, 3),
        "gaps": round(gaps, 3),
        "spread": round(spread, 3),
        "compactness": round(compactness, 3),
        "composite": round(composite, 3),
    }


# ─── Schedule Generator Service ─────────────────────────────────────────────
class ScheduleGeneratorService:
    """Service to generate all possible schedule combinations."""

    def __init__(self, db: Session):
        self.db = db
        self.course_repo = CourseRepository(db)
        self.section_repo = SectionRepository(db)
        self.session_repo = SessionRepository(db)

    # ── Public API ───────────────────────────────────────────────────────

    def generate_schedule_combinations_from_selected_sections(
        self,
        selected_section_ids: List[int],
        optional_section_ids: List[int] | None = None,
        max_optional_courses: int | None = None,
        sort_by: str = "score",
        max_results: int = 200,
    ) -> List[Dict]:
        """
        Generate combinations from user-selected sections.

        Args:
            selected_section_ids: Sections from REQUIRED courses (must all appear).
            optional_section_ids: Sections from OPTIONAL courses (pick N of M).
            max_optional_courses: How many optional courses to include (None = all).
            sort_by: "score" | "earliest" | "compact" | "spread" | "gaps"
            max_results: Cap on returned combinations.
        """
        required_options = self._load_course_options(selected_section_ids)
        optional_options = self._load_course_options(optional_section_ids or [])

        if not required_options and not optional_options:
            return []

        # Generate required base combinations
        required_combos = self._generate_valid_combinations(required_options)

        if not optional_options:
            # No optionals → score, sort, return
            scored = [
                self._combo_to_dict(opts, score_combination(opts))
                for opts in required_combos
            ]
            return self._sort_and_limit(scored, sort_by, max_results)

        # With optionals: for each required base, try adding optional combos
        optional_groups = self._group_by_course(optional_options)
        optional_course_codes = list(optional_groups.keys())

        # Determine how many optional courses to pick
        pick_count = max_optional_courses or len(optional_course_codes)
        pick_count = min(pick_count, len(optional_course_codes))

        all_results = []

        for base_opts in required_combos:
            base_bitmap = TimeSlotBitmap()
            bitmap_ok = True
            for opt in base_opts:
                if not opt.add_to_bitmap(base_bitmap):
                    bitmap_ok = False
                    break
            if not bitmap_ok:
                continue

            # Try each subset of optional courses
            for opt_course_subset in itertools.combinations(optional_course_codes, pick_count):
                opt_section_lists = [optional_groups[code] for code in opt_course_subset]

                for opt_section_combo in itertools.product(*opt_section_lists):
                    # Check if optional sections fit with the required base
                    test_bitmap = base_bitmap.copy()
                    valid = True

                    for opt_section in opt_section_combo:
                        if not opt_section.add_to_bitmap(test_bitmap):
                            valid = False
                            break

                    if valid:
                        full_opts = list(base_opts) + list(opt_section_combo)
                        scores = score_combination(full_opts)
                        all_results.append(self._combo_to_dict(full_opts, scores))

                        if len(all_results) >= max_results * 2:
                            # Early exit to avoid excessive computation
                            break

        return self._sort_and_limit(all_results, sort_by, max_results)

    # ── Internal Methods ─────────────────────────────────────────────────

    def _load_course_options(self, section_ids: List[int]) -> List[CourseOption]:
        """Load CourseOption objects from section IDs."""
        options = []
        for section_id in section_ids:
            section = self.section_repo.get(section_id)
            if not section:
                continue
            course = self.course_repo.get(section.course_id)
            if not course:
                continue
            sessions = self.session_repo.get_by_section(section_id)
            if sessions:
                options.append(CourseOption(section, sessions, course))
        return options

    def _group_by_course(self, options: List[CourseOption]) -> Dict[str, List[CourseOption]]:
        """Group options by course code."""
        groups: Dict[str, List[CourseOption]] = {}
        for opt in options:
            groups.setdefault(opt.course_code, []).append(opt)
        return groups

    def _generate_valid_combinations(
        self, options: List[CourseOption]
    ) -> List[List[CourseOption]]:
        """
        Generate all valid (no-collision) combinations using Cartesian product
        with incremental bitmap validation for early pruning.
        """
        groups = self._group_by_course(options)
        course_codes = list(groups.keys())

        if not course_codes:
            return []

        # Use recursive backtracking with bitmap pruning
        results: List[List[CourseOption]] = []
        self._backtrack(groups, course_codes, 0, [], TimeSlotBitmap(), results, max_results=500)
        return results

    def _backtrack(
        self,
        groups: Dict[str, List[CourseOption]],
        course_codes: List[str],
        depth: int,
        current: List[CourseOption],
        bitmap: TimeSlotBitmap,
        results: List[List[CourseOption]],
        max_results: int,
    ):
        """Recursive backtracking with bitmap-based pruning."""
        if len(results) >= max_results:
            return

        if depth == len(course_codes):
            results.append(list(current))
            return

        code = course_codes[depth]
        for option in groups[code]:
            # Try adding this option to the bitmap
            new_bitmap = bitmap.copy()
            if option.add_to_bitmap(new_bitmap):
                current.append(option)
                self._backtrack(
                    groups, course_codes, depth + 1,
                    current, new_bitmap, results, max_results,
                )
                current.pop()

    def _combo_to_dict(self, options: List[CourseOption], scores: Dict[str, float]) -> Dict:
        """Convert a list of CourseOptions + scores to the API response dict."""
        sorted_opts = sorted(options, key=lambda x: x.section_id)
        return {
            "combination_id": str(uuid.uuid4()),
            "id_string": "_".join(str(o.section_id) for o in sorted_opts),
            "course_count": len(sorted_opts),
            "scores": scores,
            "courses": [
                {
                    "course_id": opt.course_id,
                    "course_code": opt.course_code,
                    "course_name": opt.course_name,
                    "section_id": opt.section_id,
                    "section_number": opt.section_number,
                    "professor": opt.professor,
                    "sessions": [
                        {
                            "session_id": session.id,
                            "session_type": session.session_type,
                            "day": session.day,
                            "start_time": session.start_time.strftime("%H:%M"),
                            "end_time": session.end_time.strftime("%H:%M"),
                            "location": session.location,
                            "modality": session.modality,
                        }
                        for session in opt.sessions
                    ],
                }
                for opt in sorted_opts
            ],
        }

    def _sort_and_limit(
        self, combos: List[Dict], sort_by: str, max_results: int
    ) -> List[Dict]:
        """Sort combinations by the requested metric and truncate."""
        key_map = {
            "score": lambda c: c.get("scores", {}).get("composite", 0),
            "earliest": lambda c: c.get("scores", {}).get("earliest", 0),
            "gaps": lambda c: c.get("scores", {}).get("gaps", 0),
            "spread": lambda c: c.get("scores", {}).get("spread", 0),
            "compact": lambda c: c.get("scores", {}).get("compactness", 0),
        }
        sort_fn = key_map.get(sort_by, key_map["score"])
        combos.sort(key=sort_fn, reverse=True)
        return combos[:max_results]

File: frontend/src/features/schedule/logic/scheduleGenerator.ts
/**
 * Schedule Generator (pure TypeScript port of backend/app/services/schedule_generator.py)
 *
 * IMPORTANT decoupling change vs. the Python version: the original service loads
 * sections from the DB (`_load_course_options`). This module does NO I/O — callers
 * must pass already-loaded `SectionOption[]` in the input. No React, no supabase.
 */

// ─── Types ───────────────────────────────────────────────────────────────────

export type SortBy = "score" | "earliest" | "compact" | "spread" | "gaps";

export interface SessionOption {
  /** Session id (mirrors backend Session.id). */
  id: number;
  session_type?: string | null;
  /** Day as a string ("lunes", "monday", ...) or numeric 0-6 / 1-7. */
  day: string | number | null;
  /** "HH:MM" 24h format. */
  start_time: string;
  end_time: string;
  location?: string | null;
  modality?: string | null;
}

export interface SectionOption {
  section_id: number;
  course_id: number;
  course_code: string;
  course_name: string;
  section_number: string;
  professor?: string | null;
  /** Present on the section for reference; not used by the scoring algorithm. */
  capacity?: number | null;
  enrolled?: number | null;
  sessions: SessionOption[];
}

export interface GenerateCombinationsInput {
  sections: SectionOption[];
  /** Section ids from REQUIRED courses (must all appear in every combination). */
  selectedSectionIds: number[];
  /** Section ids from OPTIONAL courses (pick N of M). */
  optionalSectionIds?: number[];
  /** How many optional courses to include (undefined/null = all). */
  maxOptionalCourses?: number | null;
  sortBy?: SortBy;
  maxResults?: number;
}

export interface CombinationScores {
  earliest: number;
  gaps: number;
  spread: number;
  compactness: number;
  composite: number;
}

export interface CombinationSessionDict {
  session_id: number;
  session_type: string | null | undefined;
  day: string | number | null;
  start_time: string;
  end_time: string;
  location: string | null | undefined;
  modality: string | null | undefined;
}

export interface CombinationCourseDict {
  course_id: number;
  course_code: string;
  course_name: string;
  section_id: number;
  section_number: string;
  professor: string | null | undefined;
  sessions: CombinationSessionDict[];
}

export interface CombinationDict {
  combination_id: string;
  id_string: string;
  course_count: number;
  scores: CombinationScores;
  courses: CombinationCourseDict[];
}

// ─── Constants ───────────────────────────────────────────────────────────────

const DAY_ORDER: Record<string, number> = {
  // Spanish
  lunes: 0, martes: 1, "miércoles": 2, miercoles: 2,
  jueves: 3, viernes: 4, "sábado": 5, sabado: 5, domingo: 6,
  // English
  monday: 0, tuesday: 1, wednesday: 2,
  thursday: 3, friday: 4, saturday: 5, sunday: 6,
};

function parseDayToIdx(dayVal: string | number | null | undefined): number {
  if (dayVal === null || dayVal === undefined) return -1;

  if (typeof dayVal === "number") {
    if (dayVal >= 1 && dayVal <= 7) return dayVal - 1;
    if (dayVal >= 0 && dayVal <= 6) return dayVal;
    return -1;
  }

  const dayStr = String(dayVal).trim().toLowerCase();
  if (!dayStr) return -1;

  if (/^\d+$/.test(dayStr)) {
    const val = parseInt(dayStr, 10);
    if (val >= 1 && val <= 7) return val - 1;
    if (val >= 0 && val <= 6) return val;
    return -1;
  }

  if (dayStr in DAY_ORDER) return DAY_ORDER[dayStr];

  const prefix = dayStr.slice(0, 3);
  for (const key of Object.keys(DAY_ORDER)) {
    if (key.startsWith(prefix)) return DAY_ORDER[key];
  }

  return -1;
}

const SCORE_WEIGHTS = {
  earliest: 0.3,
  gaps: 0.3,
  spread: 0.25,
  compactness: 0.15,
};

const MAX_MINUTES = 24 * 60; // normalizer

// ─── Time Slot Bitmap ────────────────────────────────────────────────────────

/**
 * Set-based time-slot tracker for O(1) collision detection.
 * Each day has slots in 5-minute granularity (288 slots/day), keyed as
 * `day * 288 + slot`.
 */
class TimeSlotBitmap {
  static readonly SLOT_SIZE = 5; // minutes
  static readonly SLOTS_PER_DAY = 288;

  private occupied: Set<number> = new Set();

  copy(): TimeSlotBitmap {
    const next = new TimeSlotBitmap();
    next.occupied = new Set(this.occupied);
    return next;
  }

  /** Try to add a session. Returns false (no mutation) on collision. */
  addSession(dayIdx: number, startMinutes: number, endMinutes: number): boolean {
    if (dayIdx < 0 || dayIdx > 6) return true; // unknown day -> no collision

    const slotsNeeded: number[] = [];
    for (let t = startMinutes; t < endMinutes; t += TimeSlotBitmap.SLOT_SIZE) {
      const slot = Math.floor(t / TimeSlotBitmap.SLOT_SIZE);
      const key = dayIdx * TimeSlotBitmap.SLOTS_PER_DAY + slot;
      if (this.occupied.has(key)) return false;
      slotsNeeded.push(key);
    }

    for (const key of slotsNeeded) this.occupied.add(key);
    return true;
  }

  canAddSession(dayIdx: number, startMinutes: number, endMinutes: number): boolean {
    if (dayIdx < 0 || dayIdx > 6) return true;

    for (let t = startMinutes; t < endMinutes; t += TimeSlotBitmap.SLOT_SIZE) {
      const slot = Math.floor(t / TimeSlotBitmap.SLOT_SIZE);
      const key = dayIdx * TimeSlotBitmap.SLOTS_PER_DAY + slot;
      if (this.occupied.has(key)) return false;
    }
    return true;
  }
}

// ─── Course Option ───────────────────────────────────────────────────────────

function parseHHMM(value: string): { hour: number; minute: number } {
  const [h, m] = value.split(":").map((v) => parseInt(v, 10));
  return { hour: h || 0, minute: m || 0 };
}

/** Represents a course section with its sessions, pre-computed for speed. */
class CourseOption {
  readonly sectionId: number;
  readonly courseId: number;
  readonly courseCode: string;
  readonly courseName: string;
  readonly sectionNumber: string;
  readonly professor: string | null | undefined;
  readonly sessions: SessionOption[];
  private readonly timeSlots: Array<[number, number, number]>;

  constructor(section: SectionOption) {
    this.sectionId = section.section_id;
    this.courseId = section.course_id;
    this.courseCode = section.course_code;
    this.courseName = section.course_name;
    this.sectionNumber = section.section_number;
    this.professor = section.professor;
    this.sessions = section.sessions;

    this.timeSlots = section.sessions.map((s) => {
      const dayIdx = parseDayToIdx(s.day);
      const start = parseHHMM(s.start_time);
      const end = parseHHMM(s.end_time);
      return [
        dayIdx,
        start.hour * 60 + start.minute,
        end.hour * 60 + end.minute,
      ] as [number, number, number];
    });
  }

  get timeSlotsList(): Array<[number, number, number]> {
    return this.timeSlots;
  }

  fitsBitmap(bitmap: TimeSlotBitmap): boolean {
    for (const [dayIdx, start, end] of this.timeSlots) {
      if (!bitmap.canAddSession(dayIdx, start, end)) return false;
    }
    return true;
  }

  /** Add this option's sessions to bitmap. Returns false on collision. */
  addToBitmap(bitmap: TimeSlotBitmap): boolean {
    for (const [dayIdx, start, end] of this.timeSlots) {
      if (!bitmap.addSession(dayIdx, start, end)) return false;
    }
    return true;
  }
}

// ─── Scoring ─────────────────────────────────────────────────────────────────

/**
 * Round to 3 decimals using round-half-to-even (Python's `round(x, 3)`
 * semantics), operating on the double's exact decimal expansion via
 * `toFixed(20)` so we don't fabricate false ties from binary floating-point
 * noise the way naive `Math.round(x * 1000) / 1000` does.
 */
function round3(x: number): number {
  if (!Number.isFinite(x)) return x;
  const sign = x < 0 ? -1 : 1;
  const abs = Math.abs(x);
  const str = abs.toFixed(20);
  const [intPart, fracPart] = str.split(".");
  const kept = fracPart.slice(0, 3);
  const rest = fracPart.slice(3);
  let n = BigInt(intPart + kept);

  const first = rest[0];
  let roundUp: boolean;
  if (first < "5") {
    roundUp = false;
  } else if (first > "5") {
    roundUp = true;
  } else {
    const remainder = rest.slice(1).replace(/0+$/, "");
    roundUp = remainder.length > 0 ? true : n % BigInt(2) === BigInt(1);
  }
  if (roundUp) n += BigInt(1);

  return (Number(n) / 1000) * sign;
}

function scoreCombination(options: CourseOption[]): CombinationScores {
  if (options.length === 0) {
    return { earliest: 0, gaps: 0, spread: 0, compactness: 0, composite: 0 };
  }

  const daySessions = new Map<number, Array<[number, number]>>();
  const allStarts: number[] = [];

  for (const opt of options) {
    for (const [dayIdx, start, end] of opt.timeSlotsList) {
      if (dayIdx >= 0) {
        if (!daySessions.has(dayIdx)) daySessions.set(dayIdx, []);
        daySessions.get(dayIdx)!.push([start, end]);
        allStarts.push(start);
      }
    }
  }

  if (allStarts.length === 0) {
    return { earliest: 0, gaps: 0, spread: 0, compactness: 0, composite: 0 };
  }

  // 1. Earliest score: lower average start = better
  const avgStart = allStarts.reduce((a, b) => a + b, 0) / allStarts.length;
  const earliest = 1.0 - avgStart / MAX_MINUTES;

  // 2. Gap score: fewer minutes of gaps between consecutive classes = better
  let totalGapMinutes = 0;
  for (const intervals of daySessions.values()) {
    const sorted = [...intervals].sort((a, b) => a[0] - b[0]);
    for (let i = 1; i < sorted.length; i++) {
      const gap = sorted[i][0] - sorted[i - 1][1];
      if (gap > 0) totalGapMinutes += gap;
    }
  }
  const gaps = Math.max(0.0, 1.0 - totalGapMinutes / 300.0);

  // 3. Spread score: how evenly distributed across days
  const activeDays = daySessions.size;
  const totalSessions = [...daySessions.values()].reduce((a, v) => a + v.length, 0);

  let spread: number;
  if (activeDays > 0 && totalSessions > 0) {
    const sessionsPerDay: number[] = [];
    for (let d = 0; d < 7; d++) sessionsPerDay.push(daySessions.get(d)?.length ?? 0);
    const activeCounts = sessionsPerDay.filter((c) => c > 0);
    if (activeCounts.length > 1) {
      const avgPerDay = totalSessions / activeDays;
      const variance =
        activeCounts.reduce((sum, c) => sum + (c - avgPerDay) ** 2, 0) / activeCounts.length;
      spread = Math.max(0.0, 1.0 - variance / 4.0);
    } else {
      spread = 0.0; // all on one day = bad spread
    }
  } else {
    spread = 0.0;
  }

  // 4. Compactness: fewer active days = better (less commute)
  const compactness = Math.max(0.0, 1.0 - (activeDays - 1) / 5.0);

  const composite =
    SCORE_WEIGHTS.earliest * earliest +
    SCORE_WEIGHTS.gaps * gaps +
    SCORE_WEIGHTS.spread * spread +
    SCORE_WEIGHTS.compactness * compactness;

  return {
    earliest: round3(earliest),
    gaps: round3(gaps),
    spread: round3(spread),
    compactness: round3(compactness),
    composite: round3(composite),
  };
}

// ─── Generation ──────────────────────────────────────────────────────────────

function groupByCourse(options: CourseOption[]): Map<string, CourseOption[]> {
  const groups = new Map<string, CourseOption[]>();
  for (const opt of options) {
    if (!groups.has(opt.courseCode)) groups.set(opt.courseCode, []);
    groups.get(opt.courseCode)!.push(opt);
  }
  return groups;
}

function backtrack(
  groups: Map<string, CourseOption[]>,
  courseCodes: string[],
  depth: number,
  current: CourseOption[],
  bitmap: TimeSlotBitmap,
  results: CourseOption[][],
  maxResults: number,
): void {
  if (results.length >= maxResults) return;

  if (depth === courseCodes.length) {
    results.push([...current]);
    return;
  }

  const code = courseCodes[depth];
  for (const option of groups.get(code)!) {
    const newBitmap = bitmap.copy();
    if (option.addToBitmap(newBitmap)) {
      current.push(option);
      backtrack(groups, courseCodes, depth + 1, current, newBitmap, results, maxResults);
      current.pop();
      if (results.length >= maxResults) return;
    }
  }
}

/**
 * Generate all valid (no-collision) combinations using recursive backtracking
 * with bitmap pruning. Base combination cap: 500.
 */
function generateValidCombinations(options: CourseOption[]): CourseOption[][] {
  const groups = groupByCourse(options);
  const courseCodes = [...groups.keys()];
  if (courseCodes.length === 0) return [];

  const results: CourseOption[][] = [];
  backtrack(groups, courseCodes, 0, [], new TimeSlotBitmap(), results, 500);
  return results;
}

function combinationsOf<T>(items: T[], k: number): T[][] {
  const result: T[][] = [];
  const combo: T[] = [];
  function rec(start: number) {
    if (combo.length === k) {
      result.push([...combo]);
      return;
    }
    for (let i = start; i < items.length; i++) {
      combo.push(items[i]);
      rec(i + 1);
      combo.pop();
    }
  }
  rec(0);
  return result;
}

/** Cartesian product of arrays (mirrors itertools.product(*lists)). */
function cartesianProduct<T>(lists: T[][]): T[][] {
  return lists.reduce<T[][]>(
    (acc, list) => acc.flatMap((prefix) => list.map((item) => [...prefix, item])),
    [[]],
  );
}

function comboToDict(options: CourseOption[], scores: CombinationScores): CombinationDict {
  const sortedOpts = [...options].sort((a, b) => a.sectionId - b.sectionId);
  return {
    combination_id: crypto.randomUUID(),
    id_string: sortedOpts.map((o) => String(o.sectionId)).join("_"),
    course_count: sortedOpts.length,
    scores,
    courses: sortedOpts.map((opt) => ({
      course_id: opt.courseId,
      course_code: opt.courseCode,
      course_name: opt.courseName,
      section_id: opt.sectionId,
      section_number: opt.sectionNumber,
      professor: opt.professor,
      sessions: opt.sessions.map((session) => ({
        session_id: session.id,
        session_type: session.session_type,
        day: session.day,
        start_time: session.start_time,
        end_time: session.end_time,
        location: session.location,
        modality: session.modality,
      })),
    })),
  };
}

const SORT_KEY_MAP: Record<string, (c: CombinationDict) => number> = {
  score: (c) => c.scores.composite ?? 0,
  earliest: (c) => c.scores.earliest ?? 0,
  gaps: (c) => c.scores.gaps ?? 0,
  spread: (c) => c.scores.spread ?? 0,
  compact: (c) => c.scores.compactness ?? 0,
};

function sortAndLimit(combos: CombinationDict[], sortBy: SortBy, maxResults: number): CombinationDict[] {
  const keyFn = SORT_KEY_MAP[sortBy] ?? SORT_KEY_MAP.score;
  // Stable sort descending (Array#sort is stable per spec in modern JS engines).
  const sorted = [...combos].sort((a, b) => keyFn(b) - keyFn(a));
  return sorted.slice(0, maxResults);
}

// ─── Public API ──────────────────────────────────────────────────────────────

/**
 * Generate combinations from user-selected sections. Pure function, no I/O:
 * pass in the already-loaded `sections` catalog plus the ids to combine.
 */
export function generateCombinations(input: GenerateCombinationsInput): CombinationDict[] {
  const {
    sections,
    selectedSectionIds,
    optionalSectionIds = [],
    maxOptionalCourses = null,
    sortBy = "score",
    maxResults = 200,
  } = input;

  const sectionsById = new Map(sections.map((s) => [s.section_id, s]));

  const loadOptions = (ids: number[]): CourseOption[] => {
    const options: CourseOption[] = [];
    for (const id of ids) {
      const section = sectionsById.get(id);
      if (!section) continue;
      if (!section.sessions || section.sessions.length === 0) continue;
      options.push(new CourseOption(section));
    }
    return options;
  };

  const requiredOptions = loadOptions(selectedSectionIds);
  const optionalOptions = loadOptions(optionalSectionIds);

  if (requiredOptions.length === 0 && optionalOptions.length === 0) return [];

  const requiredCombos = generateValidCombinations(requiredOptions);

  if (optionalOptions.length === 0) {
    const scored = requiredCombos.map((opts) => comboToDict(opts, scoreCombination(opts)));
    return sortAndLimit(scored, sortBy, maxResults);
  }

  const optionalGroups = groupByCourse(optionalOptions);
  const optionalCourseCodes = [...optionalGroups.keys()];

  let pickCount = maxOptionalCourses ?? optionalCourseCodes.length;
  pickCount = Math.min(pickCount, optionalCourseCodes.length);

  const allResults: CombinationDict[] = [];

  for (const baseOpts of requiredCombos) {
    const baseBitmap = new TimeSlotBitmap();
    let bitmapOk = true;
    for (const opt of baseOpts) {
      if (!opt.addToBitmap(baseBitmap)) {
        bitmapOk = false;
        break;
      }
    }
    if (!bitmapOk) continue;

    for (const optCourseSubset of combinationsOf(optionalCourseCodes, pickCount)) {
      const optSectionLists = optCourseSubset.map((code) => optionalGroups.get(code)!);

      // NB: mirrors the Python's bare `break`, which only exits this
      // innermost loop (product of optional sections) — not the subset or
      // base-combo loops above it. Replicated verbatim for parity, even
      // though it means the 2x cap is a soft/local guard, not a global stop.
      for (const optSectionCombo of cartesianProduct(optSectionLists)) {
        const testBitmap = baseBitmap.copy();
        let valid = true;

        for (const optSection of optSectionCombo) {
          if (!optSection.addToBitmap(testBitmap)) {
            valid = false;
            break;
          }
        }

        if (valid) {
          const fullOpts = [...baseOpts, ...optSectionCombo];
          const scores = scoreCombination(fullOpts);
          allResults.push(comboToDict(fullOpts, scores));

          if (allResults.length >= maxResults * 2) {
            break;
          }
        }
      }
    }
  }

  return sortAndLimit(allResults, sortBy, maxResults);
}
